mesos-rxjava-core/src/main/java/org/apache/mesos/rx/java/MesosSchedulerClient.java

```diff
@@ -114,4 +138,75 @@ public Subscription sink(@NotNull final Observable<SinkOperation<Send>> spout) {
             .observeOn(Rx.compute())
             .subscribe(subscriber);
     }
+
+    @VisibleForTesting
+    static final class UserAgent {
+        private static final Joiner JOINER = Joiner.on(" ").skipNulls();
+        @NotNull
+        private final List<UserAgentEntry> entries;
+
+        @SafeVarargs
+        public UserAgent(@NotNull final Function<Class<? extends UserAgent>, UserAgentEntry>... entries) {
+            this.entries =
+                Arrays.asList(entries)
+                    .stream()
+                    .map(f -> f.apply(this.getClass()))
+                    .collect(Collectors.toList());
+        }
+
+        @Override
+        public String toString() {
+            return JOINER.join(entries);
+        }
+        
+        @NotNull
+        @VisibleForTesting
+        static Function<Class<? extends UserAgent>, UserAgentEntry> userAgentEntryForGradleArtifact(@NotNull final String artifactId) {
+            return (Class<? extends UserAgent> c) -> {
+                final Properties props = loadProperties(c, String.format("/META-INF/%s.properties", artifactId));
+                return new UserAgentEntry(props.getProperty("artifactId", artifactId), props.getProperty("Implementation-Version", "unknown-version"));
+            };
+        }
+
+        @NotNull
+        @VisibleForTesting
+        static Function<Class<? extends UserAgent>, UserAgentEntry> userAgentEntryForMavenArtifact(@NotNull final String groupId, @NotNull final String artifactId) {
+            return (Class<? extends UserAgent> c) -> {
+                final Properties props = loadProperties(c, String.format("/META-INF/maven/%s/%s/pom.properties", groupId, artifactId));
+                return new UserAgentEntry(props.getProperty("artifactId", artifactId), props.getProperty("version", "unknown-version"));
+            };
+        }
+
+        @NotNull
+        private static Properties loadProperties(final Class c, final String resourcePath) {
+            final Properties props = new Properties();
+            try {
+                final InputStream resourceAsStream = c.getResourceAsStream(resourcePath);
+                if (resourceAsStream != null) {
+                    props.load(resourceAsStream);
+                    resourceAsStream.close();
+                }
+            } catch (IOException e) {
+                LOGGER.warn("Unable to load classpath resources " + resourcePath, e);
+            }
+            return props;
+        }
+        
+    }
+
+    @VisibleForTesting
+    static final class UserAgentEntry {
+        private final String name;
+        private final String version;
+
+        private UserAgentEntry(final String name, final String version) {
```

Comment:
These parameters and their associated fields should be annotated with `@NonNull`. Alternatively, we could allow the `version` parameter to be `null`, but set the field value in that case to `"unknown-version"`, as a uniform way of handling missing versions. We might want to check that `name` is non-`null` as well.